File: frontend/src/services/cctvService.ts
/**
 * MineMind AI - CCTV Intelligence Service
 * Interacts with local FastAPI CCTV endpoints with offline fallback.
 * 100% Offline & Local.
 */
import {
  CameraConfig,
  CameraFrameAnalysis,
  CCTVEvent,
  CameraHealthSummary,
  VirtualSafetyZone,
  IncidentCCTVReplay
} from '../types/cctv';

const API_BASE = 'http://localhost:8000/api/v1/cctv';

const DEFAULT_CAMERAS: CameraConfig[] = [
  {
    camera_id: 'CAM-PIT-01',
    camera_name: 'Pit Floor Shovel CAM #01',
    zone_id: 'ZONE-PIT-01',
    zone_name: 'Pit Floor Loading Zone',
    mine_id: 'MINE-ALPHA-01',
    coordinates: [150.0, -120.0, 35.0],
    orientation_deg: 45.0,
    fov_degrees: 75.0,
    resolution: '1920x1080',
    fps: 30,
    operational_status: 'ACTIVE',
    ai_monitoring_status: 'ENABLED',
    recording_status: 'RECORDING',
    health_score_pct: 99.2,
    latency_ms: 12.5,
    dropped_frames_pct: 0.02,
    last_heartbeat: new Date().toISOString(),
    ptz: { pan_deg: 10.0, tilt_deg: -20.0, zoom_level: 1.2, is_patrolling: true },
    supported_models: ['PPE_COMPLIANCE', 'RESTRICTED_ZONE_GEOFENCE'],
    stream_type: 'SYNTHETIC_SIMULATOR'
  },
  {
    camera_id: 'CAM-NW-02',
    camera_name: 'North Wall Highwall Monitor #02',
    zone_id: 'ZONE-NW-01',
    zone_name: 'North-West Sector Benches',
    mine_id: 'MINE-ALPHA-01',
    coordinates: [85.0, 180.0, 95.0],
    orientation_deg: 135.0,
    fov_degrees: 80.0,
    resolution: '2560x1440',
    fps: 25,
    operational_status: 'ACTIVE',
    ai_monitoring_status: 'ENABLED',
    recording_status: 'RECORDING',
    health_score_pct: 97.8,
    latency_ms: 18.0,
    dropped_frames_pct: 0.08,
    last_heartbeat: new Date().toISOString(),
    ptz: { pan_deg: 0.0, tilt_deg: -30.0, zoom_level: 1.5, is_patrolling: false },
    supported_models: ['RESTRICTED_ZONE_GEOFENCE'],
    stream_type: 'SYNTHETIC_SIMULATOR'
  },
  {
    camera_id: 'CAM-RAMP-03',
    camera_name: 'Haul Road Main Incline Junction #03',
    zone_id: 'ZONE-RAMP-01',
    zone_name: 'Haul Road Switchback #2',
    mine_id: 'MINE-ALPHA-01',
    coordinates: [-60.0, -40.0, 60.0],
    orientation_deg: 210.0,
    fov_degrees: 65.0,
    resolution: '1920x1080',
    fps: 30,
    operational_status: 'ACTIVE',
    ai_monitoring_status: 'ENABLED',
    recording_status: 'RECORDING',
    health_score_pct: 98.9,
    latency_ms: 14.0,
    dropped_frames_pct: 0.03,
    last_heartbeat: new Date().toISOString(),
    ptz: { pan_deg: -15.0, tilt_deg: -10.0, zoom_level: 1.0, is_patrolling: true },
    supported_models: ['VEHICLE_TRACKER', 'PROXIMITY_SAFETY'],
    stream_type: 'SYNTHETIC_SIMULATOR'
  },
  {
    camera_id: 'CAM-CRU-04',
    camera_name: 'Primary Gyratory Crusher Hopper #04',
    zone_id: 'ZONE-CRU-01',
    zone_name: 'Primary Crusher Discharge',
    mine_id: 'MINE-ALPHA-01',
    coordinates: [280.0, 210.0, 110.0],
    orientation_deg: 315.0,
    fov_degrees: 70.0,
    resolution: '1920x1080',
    fps: 30,
    operational_status: 'ACTIVE',
    ai_monitoring_status: 'ENABLED',
    recording_status: 'RECORDING',
    health_score_pct: 96.4,
    latency_ms: 16.8,
    dropped_frames_pct: 0.12,
    last_heartbeat: new Date().toISOString(),
    ptz: { pan_deg: 5.0, tilt_deg: -45.0, zoom_level: 1.8, is_patrolling: false },
    supported_models: ['SMOKE_FIRE_ANOMALY'],
    stream_type: 'SYNTHETIC_SIMULATOR'
  },
  {
    camera_id: 'CAM-SUMP-05',
    camera_name: 'Pit Sump Dewatering Substation #05',
    zone_id: 'ZONE-SUMP-01',
    zone_name: 'Pit Floor Sump & Pump House',
    mine_id: 'MINE-ALPHA-01',
    coordinates: [-110.0, -190.0, 15.0],
    orientation_deg: 60.0,
    fov_degrees: 60.0,
    resolution: '1920x1080',
    fps: 20,
    operational_status: 'ACTIVE',
    ai_monitoring_status: 'ENABLED',
    recording_status: 'RECORDING',
    health_score_pct: 95.1,
    latency_ms: 22.4,
    dropped_frames_pct: 0.15,
    last_heartbeat: new Date().toISOString(),
    ptz: { pan_deg: 0.0, tilt_deg: -15.0, zoom_level: 1.0, is_patrolling: false },
    supported_models: ['PPE_COMPLIANCE'],
    stream_type: 'SYNTHETIC_SIMULATOR'
  },
  {
    camera_id: 'CAM-STOCK-06',
    camera_name: 'ROM Stockpile Stacker / Reclaimer #06',
    zone_id: 'ZONE-STOCK-01',
    zone_name: 'High-Grade ROM Stockpile',
    mine_id: 'MINE-ALPHA-01',
    coordinates: [220.0, -80.0, 85.0],
    orientation_deg: 170.0,
    fov_degrees: 85.0,
    resolution: '1920x1080',
    fps: 30,
    operational_status: 'ACTIVE',
    ai_monitoring_status: 'ENABLED',
    recording_status: 'RECORDING',
    health_score_pct: 99.0,
    latency_ms: 11.9,
    dropped_frames_pct: 0.01,
    last_heartbeat: new Date().toISOString(),
    ptz: { pan_deg: 30.0, tilt_deg: -25.0, zoom_level: 1.1, is_patrolling: true },
    supported_models: ['PPE_COMPLIANCE', 'VEHICLE_TRACKER'],
    stream_type: 'SYNTHETIC_SIMULATOR'
  }
];

export const cctvService = {
  async getCameras(zoneId?: string, status?: string): Promise<CameraConfig[]> {
    try {
      const params = new URLSearchParams();
      if (zoneId) params.append('zone_id', zoneId);
      if (status) params.append('operational_status', status);
      const res = await fetch(`${API_BASE}/cameras?${params.toString()}`);
      if (res.ok) {
        return await res.json();
      }
    } catch {
      // Fallback
    }
    return DEFAULT_CAMERAS;
  },

  async getCamera(cameraId: string): Promise<CameraConfig> {
    try {
      const res = await fetch(`${API_BASE}/cameras/${cameraId}`);
      if (res.ok) {
        return await res.json();
      }
    } catch {
      // Fallback
    }
    return DEFAULT_CAMERAS.find((c) => c.camera_id === cameraId) || DEFAULT_CAMERAS[0];
  },

  async updatePTZ(cameraId: string, panDeg: number, tiltDeg: number, zoomLevel: number): Promise<CameraConfig> {
    try {
      const res = await fetch(`${API_BASE}/cameras/${cameraId}/ptz`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ pan_deg: panDeg, tilt_deg: tiltDeg, zoom_level: zoomLevel })
      });
      if (res.ok) {
        return await res.json();
      }
    } catch {
      // Fallback
    }
    const cam = DEFAULT_CAMERAS.find((c) => c.camera_id === cameraId) || DEFAULT_CAMERAS[0];
    return {
      ...cam,
      ptz: { ...cam.ptz, pan_deg: panDeg, tilt_deg: tiltDeg, zoom_level: zoomLevel }
    };
  },

  async updateCameraStatus(cameraId: string, operationalStatus: string, aiMonitoringStatus?: string): Promise<CameraConfig> {
    try {
      const res = await fetch(`${API_BASE}/cameras/${cameraId}/status`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ operational_status: operationalStatus, ai_monitoring_status: aiMonitoringStatus })
      });
      if (res.ok) {
        return await res.json();
      }
    } catch {
      // Fallback
    }
    const cam = DEFAULT_CAMERAS.find((c) => c.camera_id === cameraId) || DEFAULT_CAMERAS[0];
    return {
      ...cam,
      operational_status: operationalStatus as any,
      ai_monitoring_status: (aiMonitoringStatus as any) || cam.ai_monitoring_status
    };
  },

  async getCameraFrame(
    cameraId: string,
    options?: { injectViolation?: boolean; injectSmoke?: boolean; injectIncursion?: boolean }
  ): Promise<CameraFrameAnalysis> {
    try {
      const params = new URLSearchParams();
      if (options?.injectViolation) params.append('inject_violation', 'true');
      if (options?.injectSmoke) params.append('inject_smoke', 'true');
      if (options?.injectIncursion) params.append('inject_incursion', 'true');
      const res = await fetch(`${API_BASE}/cameras/${cameraId}/frame?${params.toString()}`);
      if (res.ok) {
        return await res.json();
      }
    } catch {
      // Fallback
    }

    return {
      camera_id: cameraId,
      camera_name: DEFAULT_CAMERAS.find((c) => c.camera_id === cameraId)?.camera_name || cameraId,
      timestamp: new Date().toISOString(),
      frame_width: 1920,
      frame_height: 1080,
      fps: 30,
      detections: [
        {
          class_name: 'EXCAVATOR',
          confidence: 0.96,
          box_2d: [0.25, 0.3, 0.7, 0.75],
          is_violation: false
        },
        {
          class_name: 'HAUL_TRUCK',
          confidence: 0.94,
          box_2d: [0.35, 0.05, 0.85, 0.4],
          is_violation: false
        },
        {
          class_name: 'PERSON',
          confidence: 0.92,
          box_2d: [0.6, 0.8, 0.88, 0.9],
          is_violation: options?.injectViolation || false,
          violation_reason: options?.injectViolation ? 'Missing Helmet in Shovel Radius' : undefined,
          track_id: 101
        }
      ],
      tracked_objects: [
        {
          track_id: 101,
          class_name: 'PERSON',
          box_2d: [0.6, 0.8, 0.88, 0.9],
          center: [0.85, 0.74],
          velocity: [0.01, 0.0],
          speed_mps: 1.2,
          trajectory: [[0.84, 0.74], [0.85, 0.74]],
          first_seen: 100,
          last_seen: 101,
          missed_frames: 0,
          is_violation: options?.injectViolation || false
        }
      ],
      zone_incursions: [],
      proximity_warnings: [],
      total_persons: 1,
      total_vehicles: 2,
      ppe_compliance_rate_pct: options?.injectViolation ? 50.0 : 100.0,
      hazard_detected: options?.injectViolation || options?.injectSmoke || options?.injectIncursion || false,
      hazard_description: options?.injectViolation ? 'PPE Non-Compliance Alert' : undefined,
      frame_sequence: 1
    };
  },

  async getEvents(filters?: {
    cameraId?: string;
    zoneId?: string;
    eventType?: string;
    severity?: string;
    limit?: number;
  }): Promise<CCTVEvent[]> {
    try {
      const params = new URLSearchParams();
      if (filters?.cameraId) params.append('camera_id', filters.cameraId);
      if (filters?.zoneId) params.append('zone_id', filters.zoneId);
      if (filters?.eventType) params.append('event_type', filters.eventType);
      if (filters?.severity) params.append('severity', filters.severity);
      if (filters?.limit) params.append('limit', filters.limit.toString());
      const res = await fetch(`${API_BASE}/events?${params.toString()}`);
      if (res.ok) {
        return await res.json();
      }
    } catch {
      // Fallback
    }

    return [
      {
        event_id: 'EVT-88A1',
        camera_id: 'CAM-PIT-01',
        camera_name: 'Pit Floor Shovel CAM #01',
        zone_id: 'ZONE-PIT-01',
        zone_name: 'Pit Floor Loading Zone',
        event_type: 'PPE_VIOLATION',
        severity: 'HIGH',
        timestamp: new Date(Date.now() - 480000).toISOString(),
        description: 'Worker detected within 15m radius of CAT 7495 shovel without safety helmet.',
        confidence: 0.94,
        bounding_boxes: [],
        metadata: {},
        incident_created: false
      },
      {
        event_id: 'EVT-77B2',
        camera_id: 'CAM-NW-02',
        camera_name: 'North Wall Highwall Monitor #02',
        zone_id: 'ZONE-NW-01',
        zone_name: 'North-West Sector Benches',
        event_type: 'RESTRICTED_ZONE_ENTRY',
        severity: 'CRITICAL',
        timestamp: new Date(Date.now() - 840000).toISOString(),
        description: 'Personnel incursion into Bench 1350 highwall exclusion buffer.',
        confidence: 0.91,
        bounding_boxes: [],
        metadata: {},
        incident_created: true,
        incident_id: 'INC-2026-001'
      }
    ];
  },

  async getHealthSummary(): Promise<CameraHealthSummary> {
    try {
      const res = await fetch(`${API_BASE}/health`);
      if (res.ok) {
        return await res.json();
      }
    } catch {
      // Fallback
    }
    return {
      total_cameras: 6,
      active_cameras: 6,
      offline_cameras: 0,
      degraded_cameras: 0,
      ai_monitoring_enabled: 6,
      average_health_score_pct: 97.7,
      average_latency_ms: 15.3,
      total_active_violations: 1
    };
  },

  async getZones(cameraId?: string): Promise<VirtualSafetyZone[]> {
    try {
      const params = new URLSearchParams();
      if (cameraId) params.append('camera_id', cameraId);
      const res = await fetch(`${API_BASE}/zones?${params.toString()}`);
      if (res.ok) {
        return await res.json();
      }
    } catch {
      // Fallback
    }

    return [
      {
        zone_id: 'VZ-PIT-01',
        camera_id: 'CAM-PIT-01',
        name: 'Excavator 15m Swing Hazard Zone',
        hazard_type: 'SWING_RADIUS',
        polygon_points: [[0.20, 0.20], [0.85, 0.20], [0.85, 0.80], [0.20, 0.80]],
        is_active: true,
        severity: 'CRITICAL',
        description: 'Rotating upper-carriage and bucket trajectory zone for Shovel-01'
      }
    ];
  },

  async getIncidentReplay(incidentId: string): Promise<IncidentCCTVReplay> {
    try {
      const res = await fetch(`${API_BASE}/replays/${incidentId}`);
      if (res.ok) {
        return await res.json();
      }
    } catch {
      // Fallback
    }

    return {
      incident_id: incidentId,
      camera_id: 'CAM-NW-02',
      camera_name: 'North Wall Highwall Monitor #02',
      event_type: 'RESTRICTED_ZONE_ENTRY',
      event_timestamp: new Date().toISOString(),
      severity: 'CRITICAL',
      description: 'Personnel incursion into Bench 1350 exclusion buffer.',
      total_replay_duration_sec: 22.0,
      replay_frames: [
        {
          frame_index: 0,
          time_offset_seconds: -10,
          timestamp: new Date().toISOString(),
          is_incident_climax: false,
          camera_id: 'CAM-NW-02',
          camera_name: 'North Wall Highwall Monitor #02',
          annotations: []
        },
        {
          frame_index: 5,
          time_offset_seconds: 0,
          timestamp: new Date().toISOString(),
          is_incident_climax: true,
          camera_id: 'CAM-NW-02',
          camera_name: 'North Wall Highwall Monitor #02',
          annotations: []
        }
      ]
    };
  }
};

File: python-ai/app/services/cctv_service.py
"""
MineMind AI - CCTV Intelligence & Mine Surveillance Service
Provides camera management, stream control, computer vision inference coordination,
event timeline logging, camera health monitoring, and incident candidate generation.
100% Offline and local.
"""
import uuid
import datetime
from typing import List, Dict, Optional, Any
from pydantic import BaseModel, Field

from minemind_cv.camera_registry import camera_registry, CameraConfig
from minemind_cv.zones import zone_manager, VirtualSafetyZone
from minemind_cv.pipeline import cv_pipeline, CameraFrameAnalysis, DetectionBoundingBox, ZoneIncursion
from app.services.incident_service import incident_service, IncidentCategory, IncidentSeverity


class CCTVEvent(BaseModel):
    event_id: str
    camera_id: str
    camera_name: str
    zone_id: str
    zone_name: str
    event_type: str  # PERSON_DETECTED, VEHICLE_DETECTED, PPE_VIOLATION, RESTRICTED_ZONE_ENTRY, PROXIMITY_WARNING, SMOKE_DETECTED, FIRE_INDICATION, ABNORMAL_ACTIVITY, CAMERA_OFFLINE, CAMERA_TAMPER_INDICATION
    severity: str  # LOW, MEDIUM, HIGH, CRITICAL
    timestamp: str
    description: str
    confidence: float
    bounding_boxes: List[Dict[str, Any]] = Field(default_factory=list)
    metadata: Dict[str, Any] = Field(default_factory=dict)
    incident_created: bool = False
    incident_id: Optional[str] = None


class CameraHealthSummary(BaseModel):
    total_cameras: int
    active_cameras: int
    offline_cameras: int
    degraded_cameras: int
    ai_monitoring_enabled: int
    average_health_score_pct: float
    average_latency_ms: float
    total_active_violations: int


class CCTVService:
    def __init__(self):
        self.events: List[CCTVEvent] = []
        self._seed_recent_events()

    def _seed_recent_events(self):
        now = datetime.datetime.now(datetime.timezone.utc)
        seeds = [
            {
                "camera_id": "CAM-PIT-01",
                "camera_name": "Pit Floor Shovel CAM #01",
                "zone_id": "ZONE-PIT-01",
                "zone_name": "Pit Floor Loading Zone",
                "event_type": "PPE_VIOLATION",
                "severity": "HIGH",
                "delta_min": 8,
                "confidence": 0.94,
                "description": "Worker detected within 15m radius of CAT 7495 shovel without safety helmet.",
                "boxes": [{"class_name": "PERSON", "box_2d": [0.45, 0.35, 0.68, 0.42], "is_violation": True}]
            },
            {
                "camera_id": "CAM-NW-02",
                "camera_name": "North Wall Highwall Monitor #02",
                "zone_id": "ZONE-NW-01",
                "zone_name": "North-West Sector Benches",
                "event_type": "RESTRICTED_ZONE_ENTRY",
                "severity": "CRITICAL",
                "delta_min": 14,
                "confidence": 0.91,
                "description": "Personnel incursion into Bench 1350 highwall exclusion buffer.",
                "boxes": [{"class_name": "PERSON", "box_2d": [0.22, 0.40, 0.45, 0.48], "is_violation": True}]
            },
            {
                "camera_id": "CAM-RAMP-03",
                "camera_name": "Haul Road Main Incline Junction #03",
                "zone_id": "ZONE-RAMP-01",
                "zone_name": "Haul Road Switchback #2",
                "event_type": "PROXIMITY_WARNING",
                "severity": "HIGH",
                "delta_min": 22,
                "confidence": 0.96,
                "description": "Light service vehicle #LV-102 within 18m blindspot envelope of ascending haul truck #HT-04.",
                "boxes": [{"class_name": "HAUL_TRUCK", "box_2d": [0.20, 0.25, 0.75, 0.65], "is_violation": False}]
            },
            {
                "camera_id": "CAM-CRU-04",
                "camera_name": "Primary Gyratory Crusher Hopper #04",
                "zone_id": "ZONE-CRU-01",
                "zone_name": "Primary Crusher Discharge",
                "event_type": "SMOKE_DETECTED",
                "severity": "CRITICAL",
                "delta_min": 45,
                "confidence": 0.88,
                "description": "Thermal haze and dense particulate plume near secondary lube pump skid.",
                "boxes": [{"class_name": "SMOKE_INDICATOR", "box_2d": [0.10, 0.55, 0.38, 0.82], "is_violation": True}]
            },
            {
                "camera_id": "CAM-STOCK-06",
                "camera_name": "ROM Stockpile Stacker / Reclaimer #06",
                "zone_id": "ZONE-STOCK-01",
                "zone_name": "High-Grade ROM Stockpile",
                "event_type": "PERSON_DETECTED",
                "severity": "LOW",
                "delta_min": 2,
                "confidence": 0.95,
                "description": "Authorized inspection personnel on boom walkway. PPE 100% compliant.",
                "boxes": [{"class_name": "PERSON", "box_2d": [0.60, 0.80, 0.88, 0.90], "is_violation": False}]
            }
        ]

        for s in seeds:
            ts = (now - datetime.timedelta(minutes=s["delta_min"])).isoformat()
            self.events.append(CCTVEvent(
                event_id=f"EVT-{uuid.uuid4().hex[:8].upper()}",
                camera_id=s["camera_id"],
                camera_name=s["camera_name"],
                zone_id=s["zone_id"],
                zone_name=s["zone_name"],
                event_type=s["event_type"],
                severity=s["severity"],
                timestamp=ts,
                description=s["description"],
                confidence=s["confidence"],
                bounding_boxes=s["boxes"],
                metadata={"source": "SYNTHETIC_SEED"}
            ))

    def get_cameras(self, zone_id: Optional[str] = None, status: Optional[str] = None) -> List[CameraConfig]:
        return camera_registry.list_cameras(zone_id=zone_id, operational_status=status)

    def get_camera(self, camera_id: str) -> Optional[CameraConfig]:
        return camera_registry.get_camera(camera_id)

    def update_ptz(self, camera_id: str, pan_deg: float, tilt_deg: float, zoom_level: float) -> Optional[CameraConfig]:
        return camera_registry.update_ptz(camera_id, pan_deg, tilt_deg, zoom_level)

    def process_camera_frame(
        self,
        camera_id: str,
        inject_violation: bool = False,
        inject_smoke: bool = False,
        inject_incursion: bool = False
    ) -> CameraFrameAnalysis:
        cam = camera_registry.get_camera(camera_id)
        if not cam:
            raise ValueError(f"Camera {camera_id} not found")

        analysis = cv_pipeline.process_frame(
            camera_id=camera_id,
            camera_name=cam.camera_name,
            inject_violation=inject_violation,
            inject_smoke=inject_smoke,
            inject_incursion=inject_incursion
        )

        # Log events if new hazards are detected
        if analysis.zone_incursions:
            for inc in analysis.zone_incursions:
                self.record_event(
                    camera_id=camera_id,
                    event_type="RESTRICTED_ZONE_ENTRY",
                    severity=inc.severity,
                    description=inc.description,
                    confidence=0.95,
                    bounding_boxes=[d.model_dump() for d in analysis.detections if d.is_violation],
                    metadata={"zone_id": inc.zone_id, "track_id": inc.track_id}
                )

        if analysis.ppe_compliance_rate_pct < 100.0:
            violators = [d for d in analysis.detections if d.class_name == "PERSON" and d.is_violation]
            if violators:
                self.record_event(
                    camera_id=camera_id,
                    event_type="PPE_VIOLATION",
                    severity="HIGH",
                    description=f"PPE Non-compliance detected on {cam.camera_name}: {violators[0].violation_reason or 'Missing PPE'}",
                    confidence=violators[0].confidence,
                    bounding_boxes=[v.model_dump() for v in violators],
                    metadata={"compliance_pct": analysis.ppe_compliance_rate_pct}
                )

        if inject_smoke or any(d.class_name == "SMOKE_INDICATOR" for d in analysis.detections):
            self.record_event(
                camera_id=camera_id,
                event_type="SMOKE_DETECTED",
                severity="CRITICAL",
                description=f"Thermal plume / smoke detected on {cam.camera_name}",
                confidence=0.89,
                bounding_boxes=[d.model_dump() for d in analysis.detections if d.class_name == "SMOKE_INDICATOR"],
                metadata={"hazard_type": "THERMAL_SMOKE"}
            )

        return analysis

    def record_event(
        self,
        camera_id: str,
        event_type: str,
        severity: str,
        description: str,
        confidence: float,
        bounding_boxes: Optional[List[Dict[str, Any]]] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> CCTVEvent:
        cam = camera_registry.get_camera(camera_id)
        zone_id = cam.zone_id if cam else "ZONE-UNKNOWN"
        zone_name = cam.zone_name if cam else "Unknown Zone"
        camera_name = cam.camera_name if cam else camera_id

        evt_id = f"EVT-{uuid.uuid4().hex[:8].upper()}"
        ts = datetime.datetime.now(datetime.timezone.utc).isoformat()

        # Automatic Incident candidate generation for CRITICAL and repeated HIGH events
        incident_created = False
        incident_id = None
        if severity == "CRITICAL" or event_type in ("SMOKE_DETECTED", "FIRE_INDICATION", "RESTRICTED_ZONE_ENTRY"):
            cat = IncidentCategory.NEAR_MISS.value
            if event_type in ("SMOKE_DETECTED", "FIRE_INDICATION"):
                cat = IncidentCategory.EQUIPMENT_FIRE.value
            elif event_type == "RESTRICTED_ZONE_ENTRY":
                cat = IncidentCategory.WORKER_DISTRESS.value

            inc = incident_service.create_incident(
                title=f"CCTV AI Trigger: {description}",
                category=cat,
                severity=IncidentSeverity.CRITICAL.value if severity == "CRITICAL" else IncidentSeverity.HIGH.value,
                zone_name=zone_name,
                reported_by=f"CCTV-AI-{camera_id}"
            )
            incident_created = True
            incident_id = inc.get("id")

        event = CCTVEvent(
            event_id=evt_id,
            camera_id=camera_id,
            camera_name=camera_name,
            zone_id=zone_id,
            zone_name=zone_name,
            event_type=event_type,
            severity=severity,
            timestamp=ts,
            description=description,
            confidence=confidence,
            bounding_boxes=bounding_boxes or [],
            metadata=metadata or {},
            incident_created=incident_created,
            incident_id=incident_id
        )

        self.events.insert(0, event)
        if len(self.events) > 500:
            self.events = self.events[:500]

        return event

    def get_events(
        self,
        camera_id: Optional[str] = None,
        zone_id: Optional[str] = None,
        event_type: Optional[str] = None,
        severity: Optional[str] = None,
        limit: int = 50
    ) -> List[CCTVEvent]:
        res = self.events
        if camera_id:
            res = [e for e in res if e.camera_id.lower() == camera_id.lower()]
        if zone_id:
            res = [e for e in res if e.zone_id.lower() == zone_id.lower()]
        if event_type:
            res = [e for e in res if e.event_type.lower() == event_type.lower()]
        if severity:
            res = [e for e in res if e.severity.lower() == severity.lower()]
        return res[:limit]

    def get_health_summary(self) -> CameraHealthSummary:
        cams = camera_registry.list_cameras()
        active = sum(1 for c in cams if c.operational_status == "ACTIVE")
        offline = sum(1 for c in cams if c.operational_status == "OFFLINE")
        degraded = sum(1 for c in cams if c.operational_status == "DEGRADED")
        ai_enabled = sum(1 for c in cams if c.ai_monitoring_status == "ENABLED")

        avg_health = sum(c.health_score_pct for c in cams) / len(cams) if cams else 0.0
        avg_latency = sum(c.latency_ms for c in cams) / len(cams) if cams else 0.0

        # Count active unacknowledged critical/high events in last 30 minutes
        recent_violations = sum(1 for e in self.events[:20] if e.severity in ("CRITICAL", "HIGH"))

        return CameraHealthSummary(
            total_cameras=len(cams),
            active_cameras=active,
            offline_cameras=offline,
            degraded_cameras=degraded,
            ai_monitoring_enabled=ai_enabled,
            average_health_score_pct=round(avg_health, 1),
            average_latency_ms=round(avg_latency, 1),
            total_active_violations=recent_violations
        )

    def get_incident_replay(self, incident_id: str) -> Dict[str, Any]:
        """
        Retrieves visual replay metadata and snapshot frames around an incident event.
        """
        matching_events = [e for e in self.events if e.incident_id == incident_id]
        if not matching_events:
            # Fallback to most recent critical event
            matching_events = [e for e in self.events if e.severity == "CRITICAL"]

        ref_event = matching_events[0] if matching_events else self.events[0] if self.events else None
        if not ref_event:
            return {"status": "NO_RECORDINGS", "incident_id": incident_id, "frames": []}

        # Generate simulated chronological replay frames (5 pre-event frames, event frame, 5 post-event frames)
        frames = []
        for i in range(-5, 6):
            sec_offset = i * 2.0
            frames.append({
                "frame_index": i + 5,
                "time_offset_seconds": sec_offset,
                "timestamp": ref_event.timestamp,
                "is_incident_climax": (i == 0),
                "camera_id": ref_event.camera_id,
                "camera_name": ref_event.camera_name,
                "annotations": ref_event.bounding_boxes if i >= 0 else []
            })

        return {
            "incident_id": incident_id,
            "camera_id": ref_event.camera_id,
            "camera_name": ref_event.camera_name,
            "event_type": ref_event.event_type,
            "event_timestamp": ref_event.timestamp,
            "severity": ref_event.severity,
            "description": ref_event.description,
            "total_replay_duration_sec": 22.0,
            "replay_frames": frames
        }


cctv_service = CCTVService()
